src/corecel/random/engine/RanluxppRngEngine.hh

```diff
@@ -0,0 +1,191 @@
+    CELER_FUNCTION RanluxppUInt intRndm64()
+    {
+        // draw two 48-bit words, but take only their low 32 bits each
+        RanluxppUInt lo = this->nextRandomBits() & 0xFFFFFFFFu;
+        RanluxppUInt hi = this->nextRandomBits() & 0xFFFFFFFFu;
+        return (lo << 32) | hi;
+    }
+
+    //! Produce the next block of random bits.
+    CELER_FUNCTION void advance()
+    {
+        RanluxppArray9 lcg = celeritas::detail::to_lcg(state_->value);
+        lcg = celeritas::detail::compute_mod_multiply(params_.state_2048, lcg);
+        state_->value = celeritas::detail::to_ranlux(lcg);
+        state_->position = 0;
+    }
+
+    /// DATA ///
+    static constexpr int offset_ = 48;
+    ParamsRef const& params_;
+    RanluxppRngState* state_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Skip `n` random numbers without generating them.
+ */
+CELER_FUNCTION void RanluxppRngEngine::skip(RanluxppUInt n)
+{
+    CELER_ASSERT(n > 0);
+    CELER_ASSERT(params_.max_position > 0);
+
+    int left = (params_.max_position - state_->position) / offset_;
+    CELER_ASSERT(left >= 0);
+    if (n < static_cast<RanluxppUInt>(left))
+    {
+        // Just skip the next few entries in the currently
+        // available bits.
+        state_->position += n * offset_;
+        CELER_ASSERT(state_->position <= params_.max_position);
+        return;
+    }
+
+    n -= left;
+    // Need to advance and possibly skip over blocks.
+    int nPerState = params_.max_position / offset_;
+    int skip = n / nPerState;
+
+    RanluxppArray9 a_skip = celeritas::detail::compute_power_modulus(
+        params_.state_2048, skip + 1);
+
+    RanluxppArray9 lcg = celeritas::detail::to_lcg(state_->value);
+    lcg = celeritas::detail::compute_mod_multiply(a_skip, lcg);
+    state_->value = celeritas::detail::to_ranlux(lcg);
+
+    // Potentially skip numbers in the freshly generated block.
+    int remaining = n - skip * nPerState;
+    CELER_ASSERT(remaining >= 0);
+    state_->position = remaining * offset_;
+    CELER_ASSERT(state_->position <= params_.max_position);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return the next random bits, generate a new block if necessary.
+ */
+CELER_FUNCTION RanluxppUInt RanluxppRngEngine::nextRandomBits()
+{
+    if (state_->position + offset_ > params_.max_position)
+    {
+        this->advance();
+    }
+
+    int idx = state_->position / 64;
+    int offset = state_->position % 64;
+    int numBits = 64 - offset;
+
+    RanluxppUInt bits = state_->value.number[idx] >> offset;
+    if (numBits < offset_)
+    {
+        bits |= state_->value.number[idx + 1] << numBits;
+    }
+    bits &= ((RanluxppUInt(1) << offset_) - 1);
+
+    state_->position += offset_;
+    CELER_ASSERT(state_->position <= params_.max_position);
```

Comment:
So this is just a fancy way of slicing out a 48-bit integer out of a large array of integers. I think a helper class to "view" the bit state could be helpful (but maybe redundant now that it's all working).